kv: prioritize errors in UpdateStateOnRemoteRetryableErr

nvb opened the pull request:
Fixes #108853.

This commit adds logic in TxnCoordSender.UpdateStateOnRemoteRetryableErr to prioritize retryable errors based on their ErrorPriority, such that lower priority errors cannot clobber state transitions performed by higher priority errors. This ensures that if a transaction is handed a transaction aborted error and then later a transaction retry error, the latter does not cause the transaction to exit an ABORTED state.

Release note (bug fix): Fixed a bug that could cause a transaction performing multiple parallel foreign key checks to return a `concurrent txn use detected` error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109510)
<!-- Reviewable:end -->

nvb committed:
kv: prioritize errors in UpdateStateOnRemoteRetryableErr

Fixes #108853.

This commit adds logic in TxnCoordSender.UpdateStateOnRemoteRetryableErr to
prioritize retryable errors based on their ErrorPriority, such that lower
priority errors cannot clobber state transitions performed by higher priority
errors. This ensures that if a transaction is handed a transaction aborted error
and then later a transaction retry error, the latter does not cause the
transaction to exit an ABORTED state.

Release note (bug fix): Fixed a bug that could cause a transaction performing
multiple parallel foreign key checks to return a `concurrent txn use detected`
error.

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Nice sleuthing :detective:  

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/109510#-:-Nd7_O4P0LinQGHIZyKF:bruzwhc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb committed:
kv: disable merge queue in TestTxnCoordSenderRetries

Avoids test flakiness.

Epic: None
Release note: None

nvanbenschoten commented:
TFTR!

bors r+
